### ISSUE TITLE
use Q.js

### ISSUE THREAD

**Issue by aheckmann** [Automattic/mongoose#1699]

**Comment by whitecolor** [comment]
So are you going to migrate to `bluebird`?

**Comment by vkarpov15** [comment]
Nope. I don't know the next logical step ATM, because everybody has their own favorite promises library and every promises library has its own syntax, its own quirks, and its own benefits/limitations. This makes it difficult to support people specifying their own promises library.

In my personal experience I've liked Q more than any other promises library, other than perhaps mpromise, and my experiences with Bluebird have been lukewarm at best.

**Comment by benjamingr** [comment]
@vkarpov15 the answer is simple: Allow library authors to pass the promise constructor into the library (or a compatible interface) - that way anyone can use their favorite implementation with its quirks and benefits and you can stay uncoupled to any promise implementation.

**Comment by vkarpov15** [comment]
That would be great if promise constructors had compatible interfaces. Suppose you have:

`var Promise1 = require('bluebird');`
`var Promise2 = require('q');`
`var Promise3 = require('mpromise');`

For `Promise1`, you need to pass a function to the constructor, otherwise Bluebird will throw an error.

`var p = new Promise1(function(resolve, reject) {});`

For `Promise2`, you need to call `.defer()` and get the `promise` property:

`var p = Promise2.defer().promise;`

And for `Promise3` you can simply use the constructor:

`var p = new Promise3()`.

So its not quite as simple as saying `mongoose.Promise = require('q');`. FWIW, this is why I like mpromise: as far as I've seen, its the library with the least amount of library-specific quirks.

**Comment by petkaantonov** [comment]
> For `Promise1`, you need to pass a function to the constructor, otherwise Bluebird will throw an error.

Because that's [the standard](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-promise-constructor). Q implements the same standard in `Q.Promise` which the user should pass as the promise constructor. Mpromise doesn't implement any standard at all.

**Comment by benjamingr** [comment]
@vkarpov15 I'm with petka here, there is a standard and libraries like Q follow it (with `Q.Promise`) as well as other popular promise libraries (Bluebird, When, RSVP etc) and native promises.

`mpromise` is just not spec compliant. It's the outlier here You can easily create an adapter for it though.

**Comment by benjamingr** [comment]
```js
var P = require('mpromise');
function PromiseConstructor(resolver){
    var p = new P;
    try{
    resolver(p.fulfill.bind(p), p.reject.bind(p));
    } catch(e){
        p.reject(e);
    }
    return p; // also, resolving with multiple args is not allowed and is a spec violation, so there's that.
}

```

**Comment by petkaantonov** [comment]
@benjamingr If resolver is not a function the exception needs to be thrown synchronously

**Comment by petkaantonov** [comment]
@whitecolor FWIW, you can promisify mongoose in one-line with `promisifyAll(require("mongoose"))`, the methods added will also perform much faster than if mongoose constructed the promises for you through the promise constructor.

**Comment by vkarpov15** [comment]
mpromise [does implement a standard](https://promisesaplus.com/). While I do love ES6, its still just a draft of a standard rather than a standard. However, I was wrong about Q's inconsistency, so this could be implemented in theory, so thanks for pointing that out :) It will require a lot of very messy changes because the "promise constructor with executor" syntax is nasty - one of the many reasons why I like mpromise.

**Comment by whitecolor** [comment]
@petkaantonov thanks, will try it

**Comment by benjamingr** [comment]
@vkarpov15 the A+ standard just specified resolution and `then` (which, IIRC mpromise doesn't handle gracefully when assimilating a promise for multiple values - generally multiple value promises are a bad idea).

The construction specification is here https://github.com/promises-aplus/constructor-spec

**Comment by petkaantonov** [comment]
> mpromise does implement a standard. While I do love ES6, its still just a draft of a standard rather than a standard. However, I was wrong about Q's inconsistency, so this could be implemented in theory, so thanks for pointing that out :) It will require a lot of very messy changes because the "promise constructor with executor" syntax is nasty - one of the many reasons why I like mpromise.

I meant a construction standard, not any standard. And ES6 promises are nailed down.

That should not be a big deal because it's trivial to implement the deferred api on top of it:

```js
function deferredFactoryFor(PromiseConstructor) {
    return function() {
        var ret = {};
        ret.promise = new PromiseConstructor(function(resolve, reject) {
            ret.resolve = resolve;
            ret.reject = reject;
        });
        return ret;
    };
}

var bluebirdDefer = deferredFactoryFor(require("bluebird").Promise);
var QDefer = deferredFactoryFor(require("q").Promise);
var RSVPDefer = deferredFactoryFor(require("rsvp").Promise);
var WhenDefer = deferredFactoryFor(require("when").Promise);
var mdefer = ????
```

**Comment by vkarpov15** [comment]
Yeah I agree, its easy enough to add a wrapper around it. Thanks for the insights.

**Comment by tamtakoe** [comment]
+1

**Comment by joshperry** [comment]
There seems to be a larger issue in the javascript community surrounding promise libraries and libraries that expose promisified interfaces. We've conflated the functionality of providing a future value (specified in the A+ standard), and syntactic sugar for easily working with those future values.

One big problem that this creates is that, as a library author, you basically can't use anything but the naked A+ promise in your code if you're going to allow users of your library to swap out the promise implementation.

It would be nice to take a little time and perhaps think through how these two concerns could be elegantly separated.

@petkaantonov any ideas around this? Promisify has worked well for me, but it seems like there could be a better way...

**Comment by benjamingr** [comment]
@joshperry 
> There seems to be a larger issue in the javascript community surrounding promise libraries and libraries that expose promisified interfaces. We've conflated the functionality of providing a future value (specified in the A+ standard), and syntactic sugar for easily working with those future values.

I think you're looking at this the wrong way - you can expose whatever interface you'd like as long as it conforms to promises/A+ it will be easy to consume it to the outside (`then` only). 

There really isn't a huge reason to allow any library to be used - it's enough to use a fast good library like bluebird most of the time - as long as it's not something like mpromise that swallows all errors it's all good.

**Comment by joshperry** [comment]
@benjamingr

The reason that most people want a library to support their favorite promise interface is, 99 times out of 100, because they prefer/are used to the syntactical sugar that library adds to the base A+ interface. I _have_ seen performance used as an argument by some users to get their favorite library chosen, but it's usually just a red herring tactic to get what they really want: the interface of their favorite promise library.

This syntactical sugar is tightly bound to the A+ interface of every promises library, which makes it very difficult for a library author to expose a promise-based interface, _and_ support all of their user's favorite promise library sugar.

There are ways around this, but the ones I've seen have been brittle (promisifyAll) or are just so messy it's not worth doing (manually wrapping the promises returned by a library).

A lot of these discussions would go away if there was an elegant convention to separate these concerns, to allow both library authors, and library consumers to use their preferred syntactical sugar. I'm not convinced that having a promise library convert a CPS interface is the best way to accomplish this. 

This all reminds me of libraries written in C++ but that expose C bindings because the C++ ABI was never standardized enough to allow other languages to indiscriminately import C++ DLLs.

**Comment by benjamingr** [comment]
> There are ways around this, but the ones I've seen have been brittle (promisifyAll) or are just so messy it's not worth doing (manually wrapping the promises returned by a library).

Why is promisifyAll brittle? I've found it rock solid and working in wrapping every library I want. If there is a case where it isn't working there's always `Promise.fromNode`. 

To be fair - I don't see the huge value in allowing swapping of the promise implementation - I'd just pick the fastest one that's easiest to work with or not at all and do wrapping - see my comments here: https://github.com/Automattic/mongoose/issues/2688#issuecomment-104422889

**Comment by joshperry** [comment]
@benjamingr 

Re: `promisifyAll`; it wasn't until the latest version that it could promisify mongoose properly without going through a bunch of machinations promisifying each schema and its prototype, and even then it didn't handle promisifying the query chaining at all. There are bound to be patterns in the future that `promisifyAll` will not work with.

You may not see the value because the promise library you use (which is actually the one I prefer as well) is currently the performance leader and has a list of excellent features, but that could change. And then what, everyone change to the then fastest library? And really, the syntactical additions do not a fast promise library make. In fact, it's very possible that native promises will outstrip even Bluebird in speed if the implementers do a good job with the native impl.

Finally, I hope that we can agree that "easiest to work with" is incredibly subjective and doesn't really provide any value in the discussion on what we're trying to accomplish here.

**Comment by vkarpov15** [comment]
My 2 cents - if you really care about shaving a couple microseconds of CPU time by using bluebird vs RSVP or something, you're not going to be using promises anyway. Frankly, if that level of optimization matters to you, you probably shouldn't be using node at all. Promises are all about syntactic sugar and writing code that's more durable and easier to maintain. Of course that is a function of which promises library the user has most experience with, which is why I see the value of letting people specify their favorite promises library.

**Comment by benjamingr** [comment]
> My 2 cents - if you really care about shaving a couple microseconds of CPU time by using bluebird vs RSVP or something, you're not going to be using promises anyway.

Why? There is very little if at all overhead. There are actual real world workloads that are impossible to handle with mpromise but are possible (and done with bluebird). If you have 5000 connections to the client and a flow chain of a client's request is `5` thens and 5 queries that's 50000 promises you're going to have at once. For bluebird 50K promises is a walk in the park - for Q or mpromise the process will terminate.

**Comment by mgcrea** [comment]
Can someone ELI5 why mongoose 4 is using an exotic (thus probably quite slow) promise implementation over native ones or bluebird (clearly the best pick for the past couple of years)? Why not allow every end user to choose it's library as long as it comply with the A+ spec? something like `require('mongoose').usePromise(require('bluebird'))`?

**Comment by benjamingr** [comment]
@mgcrea for one thing you're using free software ;)

I'm sure pull requests are welcome - and as mentioned before using bluebird with Mongoose is already a one-liner

**Comment by mgcrea** [comment]
@benjamingr, are you referring to `promisifyAll(require("mongoose"))`?
It's not what I would call a swappable implementation since you have to swap all your methods names, from `Model.find().then` with mpromise to `Model.findAsync().then` with bluebird+promisify.

If there is consensus towards removing the tightly coupling between mongoose and mpromise, I would gladly spend some time to craft a potential PR, but for now, I need a better understanding of the reasoning behind the mpromise pick before acting and wasting everybody's time.

**Comment by benjamingr** [comment]
@mgcrea 
> I need a better understanding of the reasoning behind the mpromise pick before acting and wasting everybody's time.

People asked for promises support and Mongoose is quite a mature library now over 4 years old- mpromise was chosen because it looked like a good way to add promise support at a time. There was no bluebird when the choice was made and there certainly were no native promises.

**Comment by mgcrea** [comment]
@benjamingr I though mpromise was picked recently, but it looks like it just only really spun out of mongoose core as a separate library. Thanks for the explanation.

**Comment by vkarpov15** [comment]
@mgcrea that's the whole point of #2688. We hope to be able to support arbitrary ES6 promise libraries in the next couple months.

Also, @benjamingr is right, mpromise has been a part of mongoose for quite a long time. We kept it in as a "good enough" solution while we worked on issues we believed to be more pressing. Re-addressing promises is an immediate priority now though.
